Asset Conversion pallet

gilescope opened the pull request:
A simple automated market maker pallet.

I'd direct both non-technical discussion and discussion of possible future improvements to be hosted on the forum: https://forum.polkadot.network/t/statemint-e-swap-discussion/1498

The goal of this pallet is to provide people with a simple audited AMM pallet should a substrate chain need a basic decentralised exchange.

Commit:
Add pallet dex

Commit:
Fmt

Commit:
Add RPC endpoint

Commit:
Fix RPC

Commit:
Fix the build

Commit:
Some more fixes

Commit:
Add a method to topup pallet's account

Commit:
Add support for multi-currency into Uniques

Commit:
Fix the build

Commit:
Add [transactional] + setup() + fix balances

Commit:
Improve tests

Commit:
Fix price quotation

Commit:
Code clean up

Commit:
Validate swaps

Commit:
Fmt

Commit:
Update README

gilescope committed:
add test

gilescope committed:
mint LP assets in a different instance

gilescope committed:
remove transactional as now the default

AssetsLocal renamed to Assets

Commit:
Merge pull request #1 from paritytech/giles-more

Pool assets in separate instance

gilescope committed:
merge master

Commit:
Merge pull request #2 from paritytech/giles-more

Merge master

Commit:
Revert "Merge master"

Commit:
Merge pull request #4 from jsidorenko/revert-2-giles-more

Revert "Merge master"

gilescope committed:
Merge branch 'master' into update-master

Commit:
Merge pull request #5 from paritytech/update-master

Update master

gilescope committed:
fix tests post merge.

Commit:
Merge pull request #6 from paritytech/giles-fix-tests

fix tests post merge.

gilescope committed:
attempt to set create origin

gilescope committed:
Internally allocate lp asset id.

Commit:
Merge pull request #7 from paritytech/giles-set-create-origin

Self allocate lp token ids

gilescope committed:
Simplify

Commit:
Merge pull request #8 from paritytech/giles-set-create-origin

Simplify types a little and remove setup

gilescope committed:
Merge remote-tracking branch 'origin/master' into giles-set-create-origin

gilescope committed:
Bump to be in line

gilescope committed:
additional bumps to make compile

gilescope committed:
fix compile

Commit:
Merge pull request #9 from paritytech/giles-set-create-origin

merge master

jsidorenko committed:
Merge branch 'paritytech:master' into master

gilescope committed:
less bounds

Commit:
Merge pull request #10 from jsidorenko/giles-less-bounds

less bounds

gilescope committed:
use fungible crates

Commit:
Merge pull request #11 from jsidorenko/giles-currency-to-fungable

currency to fungable

gilescope committed:
multiasset enum

gilescope committed:
only allow native currency pairs

gilescope committed:
added slippage tests

gilescope committed:
transfer into separate method

(Also fee not set in 2 places now.)

Added test where lp and user are different users.

Commit:
Merge pull request #13 from jsidorenko/giles-multi-asset-enum

multi asset enum

Commit:
Add benchmarks + weights

Commit:
Typos

Commit:
Clean up

jsidorenko committed:
Merge pull request #14 from jsidorenko/js-add-benchmarks

Add benchmarks + weights

gilescope committed:
More tests,

split error into two because it wasn't clear which parameter.

renamed liquidity to lp_tokens_minted or lp_tokens_burned in events.

gilescope committed:
tighten up naming

Commit:
Merge pull request #15 from jsidorenko/giles-more-tests

gilescope committed:
Default, zero, square root traits not needed

Also let's not force people to be compact

Commit:
Merge pull request #16 from jsidorenko/giles-more-tests

Less bounds

gilescope committed:
add keep-alive param

gilescope committed:
add insufficient liquidity test

Commit:
Merge pull request #17 from jsidorenko/keep-alive

Commit:
Fix quote() to support u64

gilescope committed:
Avoid recording balances twice

gilescope committed:
cargo fmt

gilescope committed:
Didn't mean to change error type

gilescope committed:
temp

gilescope committed:
Less

Commit:
Rework get_amount_in/get_amount_out

Commit:
Convert other places

Commit:
Rework the last piece

jsidorenko committed:
Merge pull request #18 from jsidorenko/js-check-u64

Check u64

Commit:
Merge branch 'master' into giles-remove-balances

# Conflicts:
#	frame/dex/src/lib.rs

Commit:
Merge pull request #19 from jsidorenko/giles-remove-balances

remove balances

Commit:
Typo

Commit:
Fix benchmarks

gilescope committed:
use hash trait

Commit:
Extract a native asset check into the runtime setting

Commit:
Don't set the metadata

jsidorenko committed:
Merge pull request #22 from jsidorenko/js-extract-native-asset-check

Extract a native asset check into the runtime setting

jsidorenko committed:
Merge pull request #23 from jsidorenko/js-remove-metadata-creation

Don't set the metadata

Commit:
Merge pull request #21 from jsidorenko/giles-hash-trait

use hash trait

Commit:
Merge pull request #20 from jsidorenko/js-fix-benchmarks

Fix benchmarks

Commit:
Remove spec file

Commit:
Enable multi-assets swaps by default

Commit:
Refactor conversion into u128

jsidorenko committed:
Merge pull request #25 from jsidorenko/js-improve-u128-support

Refactor conversion into u128

jsidorenko committed:
Merge pull request #24 from jsidorenko/js-enable-multi-assets-swaps

Enable multi-assets swaps by default

Commit:
Add path param to swap_token_for_exact_tokens

Commit:
Fix typo + a bit of refactoring

Commit:
Implement path param for swap_exact_tokens_for_tokens()

Commit:
Deref

Commit:
Minor fixes

gilescope committed:
Add test with sensible scale values

Commit:
Use .windows()

Commit:
Fix benchmarks

gilescope committed:
update docs

Commit:
Merge pull request #28 from jsidorenko/giles-update-docs

update docs

Commit:
Merge pull request #27 from jsidorenko/add-realistic-test

Add test with realistic values

Commit:
Fix everything :)

Commit:
Merge branch 'master' into js-multi-swaps

# Conflicts:
#	frame/dex/src/lib.rs

Commit:
Chore

Commit:
Revert

jsidorenko committed:
Merge pull request #26 from jsidorenko/js-multi-swaps

Add path param to token swap methods

Commit:
Chore

gilescope committed:
prev way of creating sub accounts lead to collisions

Commit:
Update frame/dex/src/lib.rs

Co-authored-by: Jegor Sidorenko <5252494+jsidorenko@users.noreply.github.com>

Commit:
Update frame/dex/src/lib.rs

Co-authored-by: Jegor Sidorenko <5252494+jsidorenko@users.noreply.github.com>

Commit:
Merge pull request #29 from jsidorenko/giles-truncated

separate pool accounts

Commit:
Chore

Commit:
Fmt fix on Uniques

gilescope committed:
Merge remote-tracking branch 'jeger/master' into dex-pallet

gilescope committed:
add call_index

bring code up to date with latest master

gilescope committed:
revert readme changes

gilescope committed:
add cr

gilescope committed:
revert uniques changes

gilescope committed:
reducing noise

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2180709

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2180708

bkchr commented on primitives/rpc/Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "sp-rpc"
-version = "6.0.0"
+version = "7.0.0"
```
Why?

gilescope replied:
`sc-tracing` is using sp-rpc v7.

bkchr replied:
How does that work on master? Something seems to be wrong here or something maybe pulls the dep now from crates.io?

gilescope replied:
As we don't need the RPC crate we don't need to bump this version as we're not pulling in `sc-tracing` any more so this resolves.

bkchr commented on bin/node/executor/Cargo.toml:
```diff
@@ -39,6 +39,7 @@ pallet-sudo = { version = "4.0.0-dev", path = "../../../frame/sudo" }
 pallet-timestamp = { version = "4.0.0-dev", path = "../../../frame/timestamp" }
 pallet-treasury = { version = "4.0.0-dev", path = "../../../frame/treasury" }
 pallet-transaction-payment = { version = "4.0.0-dev", path = "../../../frame/transaction-payment" }
+pallet-dex = { version = "4.0.0-dev", path = "../../../frame/dex" }
```
Same question?

gilescope replied:
Not needed now, removed.

xlc commented on frame/dex/rpc/runtime-api/src/lib.rs:
```diff
@@ -0,0 +1,31 @@
+// This file is part of Substrate.
```
didn't we deprecated custom RPC?

joepetrowski replied:
Yeah, anything new should be behind `state_call`

gilescope replied:
Thanks for pointing that out @xlc, have removed the RPC and tested out that `state_call` works as expected.

xlc commented on frame/dex/src/lib.rs:
```diff
@@ -0,0 +1,855 @@
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Units are 10ths of a percent
+		type Fee: Get<u64>;
```
should be `Permill`

xlc commented on frame/dex/src/lib.rs:
```diff
@@ -0,0 +1,855 @@
+			amount1_min: AssetBalanceOf<T>,
+			amount2_min: AssetBalanceOf<T>,
+			mint_to: T::AccountId,
+			deadline: T::BlockNumber,
```
there is no need to have deadline as we already have transaction mortality

gilescope replied:
Thanks - it makes sense that this functionality is generic and available to every transaction.

xlc commented on frame/dex/src/lib.rs:
```diff
@@ -0,0 +1,855 @@
+		/// This actually does computation. If you need to keep using it, then make sure you cache
+		/// the value and only call this once.
+		pub fn get_pool_account(pool_id: PoolIdOf<T>) -> T::AccountId {
+			let sub = T::Hashing::hash_of(&(T::PalletId::get(), pool_id)).encode();
```
why reimplement into_sub_account

gilescope replied:
`into_sub_account_truncating` when given a `pool_id` hash manages to truncate the hash such that the pool accounts collide. Using the safer `try_into_sub_account` method returns `None` for the hash. It seems those functions focus on the first 4 bytes and disregard the rest? It's essential that the pool accounts remain separate and don't collide - I'm open to suggestions.

xlc replied:
The problem is your mock config https://github.com/paritytech/substrate/pull/12984/files#diff-093e9ce3d40b1593d70b5d0e418ec01c572a206d16ff17d4ff266eb34d9356b9R66
Make it u128.

ggwpez replied:
Additionally you can add a test in the `integrity_test` hook so that this never gets executed configured with a runtime with too short account ids.

xlc commented on frame/dex/src/lib.rs:
```diff
@@ -0,0 +1,855 @@
+			let b = u128::try_from(*b).map_err(|_| Error::<T>::UnsupportedBalanceType)?;
+			let c = u128::try_from(*c).map_err(|_| Error::<T>::UnsupportedBalanceType)?;
+
+			let result = a
```
this need to promote to U256 to avoid overflow, which is going to happen

gilescope replied:
It's promoting from u64 already. U256 might be overkill, but types are plugable so maybe we need a trait to be able to promote into a larger numeric type? (and a trait to try and get back)

xlc replied:
you have to support u128 balance type

xlc commented:
There is no good reason to support Currency & Assets, which reduces compatibility and add complexity. If you really want to implement DEX, I will suggest take a look to some of the existing implementations first.

gilescope commented:
It's no secret that the dex pallet might well be used in the statemint runtime. That runtime currently uses the Balances and Assets pallets. All the bounds are against traits so it shouldn't preclude the dex being backed by other asset registry pallets that support the `fungibles` traits.

gilescope committed:
no need for deadline (#12990)

(there's generic transaction deadline functionality already)

joepetrowski commented on bin/node/runtime/src/lib.rs:
```diff
@@ -1484,6 +1485,45 @@ impl pallet_assets::Config for Runtime {
+	type Balance = u128;
+	type AssetId = u32;
+	type Currency = Balances;
+	type CreateOrigin = AsEnsureOriginWithArg<EnsureSigned<AccountId>>;
```
This needs to be the DEX pallet, this instance is not for users to register assets.

gilescope replied:
Agreed the kitchen sink should set a good example. Fixed.

joepetrowski commented on bin/node/runtime/src/lib.rs:
```diff
@@ -1534,6 +1574,9 @@ impl pallet_uniques::Config for Runtime {
 	type CollectionId = u32;
 	type ItemId = u32;
 	type Currency = Balances;
+	type CurrencyBalance = <Self as pallet_balances::Config>::Balance;
+	type Assets = Assets;
+	type AssetId = <Self as pallet_assets::Config<Instance1>>::AssetId;
```
Why the change to the `pallet_uniques` config?

gilescope replied:
Sorry this should have been reverted. Child PR to fix that here: https://github.com/paritytech/substrate/pull/12991

gilescope replied:
Fixed

gilescope committed:
fix kitchen sink (#12991)

* fix kitchen sink

* Only the dex can mint lp_tokens

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2184268

xlc commented:
You need to check how orml-tokens, orml-currencies works. I did a workshop about it https://github.com/xlc/orml-workshop
Basically the DEX pallet shouldn't need to handle two currency implementation. It is an adapter's job to handle that.

gilescope committed:
add BenchmarkHelper for second instance (#12998)

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2186041

gilescope committed:
update mock to latest master

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2186098

gilescope commented:
@xlc I hear you. I'm hoping that XCM v3 which is "coming soon™" gets merged. When that happens I'd like to re-work the interface so that MultiAssetId is replaced with XCM v3's MultiLocation as that can represent everything in a standard way.

gilescope committed:
less indirections (#13012)

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2191615

gilescope committed:
remove dex PR's custom RPC (#13050)

* As we have state_call we don't need a custom RPC

* fix docs

gilescope committed:
no longer a need to upgrade rpc version (#13053)

gilescope committed:
Merge branch 'master' into dex-pallet

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2291953

gilescope committed:
add CallbackHadle

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2292004

gilescope committed:
quote bugfix (#13191)

quote was giving same price in both directions as we were inverting needlessly.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2297194

gilescope committed:
Merge branch 'master' into dex-pallet

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2298685

joepetrowski reviewed (commented):
A lot of documentation needs improvement.

Chralt98 commented:
> @xlc I hear you. I'm hoping that XCM v3 which is "coming soon™" gets merged. When that happens I'd like to re-work the interface so that MultiAssetId is replaced with XCM v3's MultiLocation as that can represent everything in a standard way.

So we can expect that everything works with orml_tokens and orml_currencies in the first iteration of this pallet? This would be super useful, especially because many projects depend on the orml library. pallet_asset support is fine, but the real fruition kicks in with MultiAsset pallets like from the orml library. pallet_assets differs from orml_tokens in key points as we know. I don't think that pallet_assets is meant to replace orml_tokens, right?

gilescope committed:
merging in dex specific changes due to pay by dex

gilescope committed:
update lock file

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2383777

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2383776

gilescope committed:
merging in kitchen sink changes

jsidorenko commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into dex-pallet

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2487666

gilescope committed:
Merge branch 'master' into dex-pallet

Commit:
Merge remote-tracking branch 'origin/master' into dex-pallet

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2533392

Commit:
Add get_reserves() api method

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2542083

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2542076

jsidorenko commented:
bot clean

gilescope committed:
Partial updating of the benchmarks

gilescope committed:
Merge branch 'dex-pallet' of github.com:paritytech/substrate into dex-pallet

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2551632

Commit:
Merge branch 'master' into dex-pallet

# Conflicts:
#	frame/support/src/traits/tokens/fungibles.rs

Commit:
Fix tests

Commit:
clippy

Commit:
Temp fix weights

Commit:
Fix benchmarks

Commit:
Add pool setup fee

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2571383

Commit:
Money upfront

Commit:
Address some comments

jsidorenko commented:
bot clean

Commit:
Use u128 in mock

Commit:
Fix benchmarks

Commit:
Change error message

Commit:
Update comments

Commit:
Change error names

Commit:
Implement PartialOrd for NativeOrAssetId

jsidorenko commented on frame/dex/src/lib.rs:
```diff
@@ -0,0 +1,849 @@
+			asset1: MultiAssetId<T::AssetId>,
+			asset2: MultiAssetId<T::AssetId>,
+		) -> PoolIdOf<T> {
+			if asset1 <= asset2 {
+				(asset1, asset2)
+			} else {
+				(asset2, asset1)
+			}
```
I added a manual implementation of the Ord/PartialOrd trait for NativeOrAssetId enum

jsidorenko commented on frame/dex/src/lib.rs:
```diff
@@ -0,0 +1,849 @@
+			ensure!(
+				!amount2.is_zero() && amount2 >= amount2_min_receive,
+				Error::<T>::InsufficientAmountParam2
+			);
```
updated

Commit:
add note

jsidorenko commented on bin/node/runtime/src/lib.rs:
```diff
@@ -1486,6 +1488,52 @@ impl pallet_assets::Config for Runtime {
+}
+
+impl pallet_dex::Config for Runtime {
+	type Fee = ConstU64<3>;
```
done

Commit:
Update errors

Commit:
More tests for assets sorting

Commit:
Merge branch 'master' of https://github.com/paritytech/substrate into HEAD

Commit:
".git/.scripts/commands/bench/bench.sh" pallet dev pallet_dex

Commit:
Change the way we generate pool accounts

Commit:
Improve the liquidity removal method

Commit:
Extract MintMinLiquidity to config, rework all tests

Commit:
Add comments

Commit:
Validate provided amount

Commit:
Rename to asset-conversion

jsidorenko commented:
bot bench $ pallet dev pallet_asset_conversion

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2615754

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2615740

Commit:
Validate ED

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2616232

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2616218

Commit:
Improve handling the ED related errors

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2616278

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2616292

Commit:
typos

Commit:
Try to fix benchmarks

Commit:
Merge branch 'master' into dex-pallet

Commit:
Another try

ruseinov committed:
[Enhancement] Throw an error when there are too many pallets (#13763)

* [Enhancement] Throw an error when there are too many pallets

* fix ui test

* fix PR comments

* Update frame/support/procedural/src/construct_runtime/mod.rs

Co-authored-by: Bastian Köcher <git@kchr.de>

* Update frame/support/procedural/src/construct_runtime/mod.rs

Co-authored-by: Bastian Köcher <git@kchr.de>

* ".git/.scripts/commands/fmt/fmt.sh"

---------

Co-authored-by: Bastian Köcher <git@kchr.de>
Co-authored-by: command-bot <>

Commit:
Another day, another try

Commit:
Fix benchmarks

Commit:
Expose fee related params

Commit:
Validate token's minimal amount the same way as ED

gilescope committed:
fix typo

gilescope committed:
Merge branch 'master' into dex-pallet

Commit:
Use longer path for swaps in benchmarks

gilescope committed:
Merge branch 'dex-pallet' of https://github.com/paritytech/substrate into dex-pallet

gilescope committed:
need to ref sp_std's vec.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2702528

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2702527

gilescope committed:
Remove From<u32> requirement when benchmarking

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2704330

gilescope committed:
impl BenchmarkHelper for ()

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2707748

jsidorenko commented on frame/asset-conversion/src/lib.rs:
```diff
@@ -0,0 +1,1043 @@
+		/// The type used to describe the amount of fractions converted into assets.
+		type AssetBalance: Balance;
+
+		type HigherPrecisionBalance: IntegerSquareRoot
```
Missed docs

jsidorenko commented on frame/asset-conversion/src/lib.rs:
```diff
@@ -0,0 +1,1043 @@
+	}
+
+	#[pallet::storage]
+	pub type Pools<T: Config> =
```
Missed docs

jsidorenko commented on frame/asset-conversion/src/lib.rs:
```diff
@@ -0,0 +1,1043 @@
+		},
+	}
+
+	// Your Pallet's error messages.
```
```suggestion
```

jsidorenko commented on frame/asset-conversion/src/lib.rs:
```diff
@@ -0,0 +1,1043 @@
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		PoolCreated {
```
We need to put doc for each event

jsidorenko commented on frame/asset-conversion/src/lib.rs:
```diff
@@ -0,0 +1,1043 @@
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T>
+	// where
+	// T::AssetId: From<u32>,
```
```suggestion
```

jsidorenko commented on frame/asset-conversion/src/lib.rs:
```diff
@@ -0,0 +1,1043 @@
+		/// the value and only call this once.
+		pub fn get_pool_account(pool_id: PoolIdOf<T>) -> T::AccountId {
+			T::PalletId::get().into_sub_account_truncating(pool_id)
+			// let sub = sp_io::hashing::blake2_256(&Encode::encode(&pool_id)[..]);
+			// T::PalletId::get().into_sub_account_truncating(sub)
```
```suggestion
```

gilescope committed:
only for runtime benchmarks

gilescope committed:
MultiLocation: !MaybeDisplay

Looks like we might not need this bound from initial testing.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2712552

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2712553

gilescope committed:
Update frame/asset-conversion/src/lib.rs

Co-authored-by: Jegor Sidorenko <5252494+jsidorenko@users.noreply.github.com>

gilescope committed:
Update frame/asset-conversion/src/lib.rs

Co-authored-by: Jegor Sidorenko <5252494+jsidorenko@users.noreply.github.com>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2712902

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2712903

gilescope committed:
Merge branch 'master' into dex-pallet

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2712969

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2712955

gilescope committed:
Add documentation links

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2713834

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2713820

gilescope committed:
add collision test

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2714993

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2714979

gilescope committed:
Revert "[Enhancement] Throw an error when there are too many pallets (#13763)"

This reverts commit 9b8e6e780352051029b0fdc56fcac5074bd74552.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726176

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726178

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726177

gilescope committed:
Merge branch 'master' into dex-pallet

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726231

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726230

gilescope committed:
add benchmark helper

+ doc fix

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726337

gilescope committed:
cargo fmt

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726545

Commit:
Fix adding liquidity to non-empty pool

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726977

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2726963

Commit:
Fix compilation error

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2727820

Commit:
Fix params ordering issue

gilescope committed:
additional docs

Commit:
The swap path elements should be unique

gilescope committed:
Fix account collision

gilescope committed:
Merge branch 'dex-pallet' of https://github.com/paritytech/substrate into dex-pallet

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2729864

Commit:
Validate all the pool in a swap path are unique

gilescope committed:
Merge branch 'master' into dex-pallet

Commit:
Change the way we add liquidity to empty pools

gilescope committed:
Merge branch 'master' into dex-pallet

gilescope committed:
Merge branch 'master' into dex-pallet

gilescope commented on frame/dex/src/lib.rs:
````diff
@@ -0,0 +1,849 @@
+//! curl -sS -H "Content-Type: application/json" -d \
+//! '{"id":1, "jsonrpc":"2.0", "method": "state_call", "params": ["DexApi_quote_price", "0x0101000000000000000000000011"]}' \
+//! http://localhost:9933/
+//! ```
````
all items are now documented.

gilescope committed:
Improve docs

gilescope committed:
remove unnessisary Display impl

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2740688

gilescope committed:
cargo fmt

gilescope committed:
remove unused imports

Commit:
Make api consistent

Commit:
Chore

jsidorenko commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into dex-pallet

Commit:
Touch the pool account so it could hold the pair tokens

Commit:
Check the balance before touching the pool's account

Commit:
Introduce liquidity provision fee

Commit:
Touch the pool acc one more time

sam0x17 commented on frame/asset-conversion/src/lib.rs:
```diff
@@ -0,0 +1,1187 @@
+		/// Provided amount should be greater or equal to the existential deposit/asset's minimal
+		/// amount.
+		AmountLessThanMinimal,
+		/// Reserve needs to always be greater or equal to the existential deposit/asset's minimal
```
```suggestion
		/// Reserve needs to always be greater than or equal to the existential deposit/asset's minimal
```

sam0x17 commented on frame/asset-conversion/src/lib.rs:
```diff
@@ -0,0 +1,1187 @@
+		PoolExists,
+		/// Desired amount can't be zero.
+		WrongDesiredAmount,
+		/// Provided amount should be greater or equal to the existential deposit/asset's minimal
```
```suggestion
		/// Provided amount should be greater than or equal to the existential deposit/asset's minimal
```

sam0x17 commented on frame/asset-conversion/src/lib.rs:
```diff
@@ -0,0 +1,1187 @@
+		OptimalAmountLessThanDesired,
+		/// Insufficient liquidity minted.
+		InsufficientLiquidityMinted,
+		/// Asked liquidity can't be zero.
```
```suggestion
		/// Requested liquidity can't be zero.
```

sam0x17 commented on frame/asset-conversion/src/lib.rs:
```diff
@@ -0,0 +1,1187 @@
+		/// Provide liquidity into the pool of `asset1` and `asset2`.
+		/// NOTE: an optimal amount of asset1 and asset2 will be calculated and
+		/// might be different to provided `amount1_desired`/`amount2_desired`
+		/// thus it's needed to provide the min amount you're happy to provide.
```
```suggestion
		/// thus you should provide the min amount you're happy to provide.
```

sam0x17 commented on frame/asset-conversion/src/lib.rs:
```diff
@@ -0,0 +1,1187 @@
+		}
+
+		/// Swap the exact amount of `asset1` into `asset2`.
+		/// `amount_out_min` param allows to specify the min amount of the `asset2`
```
```suggestion
		/// `amount_out_min` param allows you to specify the min amount of the `asset2`
```

sam0x17 reviewed (commented):
Looks good at a high level. Will approve once others have weighed in. Found a few doc comment/grammar tweaks :)

jsidorenko commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into dex-pallet

jsidorenko committed:
Apply suggestions from code review

Co-authored-by: Sam Johnson <sam@durosoft.com>

jsidorenko committed:
Update frame/asset-conversion/README.md

Co-authored-by: Sam Johnson <sam@durosoft.com>

Commit:
Use ContainsPair instead of the balance checker

gilescope committed:
Merge branch 'master' into dex-pallet

Commit:
Remove old Currency trait

Commit:
Add liquidity withdrawal fee

Commit:
Update docs

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2872687

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2872919

Commit:
Use 0 withdrawal fee in mock

Commit:
Rename vars

gilescope committed:
Merge branch 'master' into dex-pallet

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2883967

gilescope committed:
Merge branch 'master' into dex-pallet

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2890073

gilescope committed:
asset id not clone

gilescope committed:
fix: shadow var was being used

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2890842

gilescope committed:
correct tests

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2893706

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2893689

jsidorenko commented:
let's allow auditors to set the "audited" label :)

gilescope committed:
fix benches

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2894764

gilescope committed:
Merge branch 'master' into dex-pallet

gilescope committed:
merge master

gilescope committed:
neater

gilescope committed:
Merge branch 'master' into dex-pallet

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-dependent-cumulus
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2897564

gilescope commented:
bot merge
